cds/container/hopscotch_hashmap.h

```diff
@@ -0,0 +1,660 @@
+/*
+* concurrent_hopscotch_hash_set.h
+*
+* Created on: 23.12.2017
+* Original idea by Maurice Herlihy, Nir Shavit, Moran Tzafrir: http://people.csail.mit.edu/shanir/publications/disc2008_submission_98.pdf
+* Based on implementation by Sathya Hariesh Prakash, Royston Monteiro, Mohan Pandiyan: https://github.com/harieshsathya/Hopscotch-Hashing
+* 
+* This version modifiers: Leonid Skorospelov, Andrey Chulanov, Roman Stetskevich
+*/
+
+#ifndef HOPSCOTCHHASHMAP_H_
+#define HOPSCOTCHHASHMAP_H_
+#include <iostream>
+#include <atomic>
+#include <mutex>
+#include <cstdlib>
+
+namespace cds {
+	namespace container {
+		template<class KEY, class DATA>
+		struct hopscotch_hashmap {
+
+			typedef KEY key_type;    ///< key type
+			typedef DATA mapped_type; ///< type of value stored in the map
+			typedef std::pair<key_type const, mapped_type>   value_type;   ///< Pair type
+
+			typedef cds::atomicity::item_counter item_counter;
+			item_counter m_item_counter;
+
+		private:
+			static const int HOP_RANGE = 32;
+			static const int ADD_RANGE = 256;
+			static const int MAX_SEGMENTS = 1048576;
+			static const int MAX_TRIES = 2;
+			KEY* BUSY;
+			DATA* BUSYD;
+			struct Bucket {
+
+				static const unsigned int _empty_hop_info = 0;
+				KEY* _empty_key = NULL;
+				DATA* _empty_data = NULL;
+
+				std::atomic<unsigned int> _hop_info;
+				KEY* _key;
+				DATA* _data;
+				std::atomic<unsigned int> _lock;
+				std::atomic<unsigned int> _timestamp;
+				std::mutex lock_mutex;
+
+				Bucket() {
+					_hop_info = _empty_hop_info;
+					_lock = 0;
+					_key = _empty_key;
+					_data = _empty_data;
+					_timestamp = 0;
+				}
+
+				void lock() {
+					lock_mutex.lock();
+					while (1) {
+						if (_lock == 0) {
+							_lock = 1;
+							lock_mutex.unlock();
+							break;
+						}
+					}
+				}
+
+				void unlock() {
+					lock_mutex.lock();
+					_lock = 0;
+					lock_mutex.unlock();
+				}
+
+			};
+
+			Bucket* segments_arys;
+
+			template <typename K>
+			unsigned int calc_hash(K const& key) {
+				std::hash<K> hash_fn;
+				return hash_fn(key) % MAX_SEGMENTS;
+			}
+
+			void resize() {
+				// TODO need to find out if we need one
+			}
+
+		public:
+			static bool const c_isSorted = false; ///< whether the probe set should be ordered
+
+			hopscotch_hashmap() {
+				segments_arys = new Bucket[MAX_SEGMENTS + ADD_RANGE];
+				m_item_counter.reset();
+				BUSY = (KEY*)std::malloc(sizeof(KEY));
+			}
+
+			~hopscotch_hashmap() {
+				std::free(BUSY);
+				std::free(segments_arys);
```

Comment:
This should be `delete[]` and not `std::free` right? Am I missing something?

Reply:
@DaKellyFella, the aspect of memory allocation is a point to consider rewriting in this implementation. Probably, power-of-2-policy will be enough as a default policy, not MAX_SEGMENTS, for sure.
In general, you are right - delete should stand here.

Reply:
> Probably, power-of-2-policy will be enough as a default policy

Awesome, then you can have a mask instead of a modulus for calculating the table index.